[RFC]: Eager mode assertion ops functionalization

XuanQi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104203

This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?

XuanQi committed:
[RFC]: Eager mode functionalization

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode functionalization"




[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes (might be non-trivial since current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.


[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes (might be non-trivial since current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/functional_assertions_helper.py:
```diff
@@ -0,0 +1,103 @@
+    @classmethod
+    def get_dep_token(cls):
+        if cls._dep_token is None:
+            cls._dep_token = aten._make_dep_token()
```
hmm, I see `_make_dep_token` is in `native_functions.yaml`.

We usually try to avoid adding extra ops to `native_functions.yaml`, unless they need the full power of our codegen (autograd support, eager mode support, possible has different implement for different backends, etc).

Does `_make_dep_token` need any of that? Or is it more specific to `torch.export`. If it is, it might make more sense to register it as a custom operator (this is actually pretty easy to do, since you can register ops from python!). E.g. https://github.com/pytorch/pytorch/blob/ef285faebacec86dff5fbf8a8dca8f2672a8c257/test/test_python_dispatch.py#L331

XuanQi replied:
> We usually try to avoid adding extra ops to native_functions.yaml, unless they need the full power of our codegen

I see. This is to great to learn though. And I mostly follow @tugsbayasgalan's guidance for creating aten op.

> it might make more sense to register it as a custom operator

Can definitely play with this. One noob question is that it can create an aten level op right?

bdhirsh replied:
> One noob question is that it can create an aten level op right?

Hmm, I don't think it'll let you create an op in the "aten" namespace (you could test it out - pretty sure it'll raise an error that aten has already been defined elsewhere).

But if this op is only specific to export anyway, I think it's reasonable to put it in another namespace like "torch.ops.export_ops._make_dep_token", etc

XuanQi replied:
I feel like it doesn't necessary need to be export specific since we could also enable assertion functionalization during non-export compile? But anyway, this could be a separate discussion since renaming the op will happen in separate PR anyway.

bdhirsh commented on torch/_functorch/functional_assertions_helper.py:
```diff
@@ -0,0 +1,103 @@
+class FunctionalAssertionsHelper:
+    @staticmethod
+    def can_functionalize(needs_autograd: bool) -> bool:
+        if config.functionalize_assertion_ops:
```
one question: will we ever **not** want to functionalize asserts for export? I'm wondering if this config is meant to be a long-term option (so users can decide to or not to functionalize asserts during export), or if we'll eventually want it to always be enabled.

If we want it to be always-on but only for export, one option would be to check the flag `aot_config.is_export()`, that tells us if we're running from an `aot_export_module()` call, vs the vanilla torch.compile() codepaths.

XuanQi replied:
@tugsbayasgalan: what's your thought here? I feel like maybe for now we still want to provide a way for user to opt out assertion functionalization just for safety?

> In fact, I think we probably want this check, so nobody accidentally tries to functionalize asserts while running `torch.compile`

So you're saying for eager mode (non-export) we want to make assertion functionazation opt in based (by default no functionalization)?

tugsbayasgalan replied:
Yeah I think we still need to provide a way to opt out of functionalization assertion as the downstream users are still modifying the stack to adopt our new changes to the graph. So I feel little uneasy about turning it on unconditionally yet.

bdhirsh commented on torch/_functorch/functional_assertions_helper.py:
```diff
@@ -0,0 +1,103 @@
+
+class FunctionalAssertionsHelper:
+    @staticmethod
+    def can_functionalize(needs_autograd: bool) -> bool:
```
dumb nit: rename to `can_functionalize_asserts()` just to make it clear? we have many flavors of functionalization now 😛

bdhirsh commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -26,6 +39,38 @@ def test_functional_sym_constrain_range(self) -> None:
+
+
+@contextmanager
+def _enable_functionalization():
```
nit: `_enable_functionalized_asserts`?

bdhirsh commented on test/export/test_funtionalized_assertions.py:
```diff
@@ -26,6 +39,38 @@ def test_functional_sym_constrain_range(self) -> None:
+        with _enable_functionalization():
+            compiled = torch.compile(f, backend=my_compiler)
+            inp = torch.Tensor([3, 4, 5])
+            self.assertTrue(torch._dynamo.utils.same(compiled(inp), f(inp)))
```
This is a great E2E test. If would be great if you could also a test more directly for AOTAutograd

(1) add a similar test in `test/functorch/test_aotdispatch.py`, that calls `aot_export_module()`
(2) include an expecttest in that test, that asserts on what the graph looks like (to show that is has a functional assert call in it). There are some examples in that file, e.g. [here](https://github.com/pytorch/pytorch/blob/1a8af1503f10540c4db229ea95a72a32340a965f/test/functorch/test_aotdispatch.py#L2027)

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3167,9 +3189,12 @@ def create_aot_dispatcher_function(
+        stack.enter_context(preserve_rng_state())
+        stack.enter_context(fake_mode)
+        stack.enter_context(python_dispatcher_mode)
+        stack.enter_context(PhiloxStateTracker())
```
hmm. It would be nice if we can use the normal `with ...` syntax for all of our context managers. Why did we need `ExitStack`?

XuanQi replied:
So I think the reason I changed to use `ExitStack ` is because I want to be able to enter context for assertion functionalization programmatically instead of always on. How about I change it to something like:
```
with torch.autograd.set_multithreading_enable(...), ..., ExitStack() as stack:
    ...
    if FunctionalAssertionsHelper.can_functionalize(...):
        stack.enter(FunctionalAssertionsHelper.get_state_tracker())
```

bdhirsh replied:
ah ok - the standard pattern for doing that that I've seen throughout the codebase is:
```
maybe_enable_functionalize_asserts = nullcontext()
if FunctionalAssertionsHelper.can_functionalize(...):
    maybe_enable_functionalize_asserts = FunctionalAssertionsHelper.get_state_tracker()
...

with maybe_enable_functionalize_asserts:
```

XuanQi replied:
Sure can do that. I guess the downside of it is that it adds another level of indent though :p

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3202,6 +3227,13 @@ def convert(idx, x):
             and torch.is_grad_enabled()
         )
+        if FunctionalAssertionsHelper.can_functionalize(needs_autograd=needs_autograd):
+            stack.enter_context(FunctionalAssertionsHelper.get_state_tracker())
```
does this not work?

 ```
with FunctionalAssertionsHelper.get_state_tracker():
   ...
```

XuanQi replied:
I think the motivation here is that it only enters the scope when it can instead of always entering the scope since in current implementation: entering scope == can functionalize. How about the solution as I mentioned at https://github.com/pytorch/pytorch/pull/104203#discussion_r1244392149? Or I can have another global state to track whether it can enable asserts funcitonalization or not?

bdhirsh commented:
High level: this looks great! Left a bunch of minor feedback.

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2249,6 +2266,67 @@ def fn(p, x):
+        select: i64[] = torch.ops.aten.select.int(arg1_1, 0, 0)
+        eq: b8[] = torch.ops.aten.eq.Scalar(select, 3);  select = None
+        _make_dep_token: f32[] = torch.ops.aten._make_dep_token.default()
+        _functional_assert_async: f32[] = torch.ops.aten._functional_assert_async.msg(eq, 'test assertion error', _make_dep_token);  eq = _make_dep_token = None
```
it might be worth including two asserts in the traced function, just so you can confirm from the test that the dependency exists in the graph and is being created properly.

bdhirsh commented on torch/_functorch/functional_assertions_helper.py:
```diff
@@ -0,0 +1,135 @@
+    def create_compile_tracing_wrapper(func) -> Callable:
+        def _traced_forward(*args):
+            outs = func(*args)
+            return outs + (DepTokenStateTracker.get_dep_token(),)
```
One more thing: right now, this means that if you set `config.functionalize_asserts=True`, then we will **always** end up putting a dependency token in the graph, even if there were not actually any asserts. This is potentially wasteful: we're allocating and returning an extra tensor in the graph for no reason.

Can we avoid this? One way would be if in this function, you check whether or not we've already created a dependency token: if not, that means that we don't need to return it here, and that we won't need the extra epilogue logic.

XuanQi replied:
Good catch. I originally was trying to avoid creating dep token if there is no assertion in the graph. However, this seems a little bit hard since I want to set `num_outputs_dep_token` in output of `run_functionalized_fw_and_collect_metadata` which is before the compiling? But let me double check on it, maybe I can record the whether there is assertion in the graph or not during `run_functionalized_fw_and_collect_metadata` since seems like we run the module anyway?

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]

bdhirsh reviewed (commented):
Thanks for trying out the change to add the token when we see that there are any asserts in our graph!

After looking at how that changed played out, I don't love the extra complexity that it added (every function has to plumb the updated metadata around, and the fact that it's mutated feels very confusing: who knows which bits of the metadata are mutated at which points in the tracing logic?).

I think I'd be ok with landing the simpler version, that just always includes a dependency token. If we get any complaints from people who are using functionalized asserts and want to cut down on hot-path runtime, we can address it then.

One alternative: it actually should be possible to get the metadata right the first time, but it'll be some extra work so I'd be ok with landing the PR if you don't want to make the change immediately. The basic idea is that we want those extra assert decomps to run **during** execution of `run_functionalized_fw_and_collect_metadata()`. One way to do that: we can replace the existing meta function for `aten.assert()` with the decomps, so that the decomp actually runs when we hit an assert op. The tricky bit is that we'd only want to temporarily register the decomposition, for the duration of `run_functionalized_fw_and_collect_metadata`.

One way to do that could be with a decorator:
```
@register_metas(assert_decomp_table)
def run_functionalized_fw_and_collect_metadata():
    ...
```

Where `register_metas` creates a `torch.Library()` object and registers each decomp to it on creation, and deletes the library object when the function finishes executing (which will de-register those meta functions).

XuanQi committed:
Update on "[RFC]: Eager mode assertion ops functionalization"


This PR implements logic to functionalize assertion ops in aot_autograd. A few high level notes:
* Even it mentions eager mode in title, the implementation doesn't need to be specific to eager mode - it should work for, for example, export mode as well.
* The general implementation follows RNG op functionalization logic but with the major difference here is that we don't change the graph input but instead directly generate token during graph execution, the reason is:
  * Easier to implement.
  * Be consistent with current [functionalization logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) during export.
* If we have this logic, do we still need this [logic](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/functionalize_side_effectful_ops_pass.py#L18) anymore? I would say for now yes. The reason is because [other assertion ops](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L264C9-L264C29) will be inserted into the graph after AOT export. The potential ways to resolve this is to:
  * Move adding runtime assertion before AOT export and purely rely on assertion functionalization logic in aot_autograd. Note that this requires extra changes. This might not be a trivial change since:
    * Current `_AddRuntimeAssertionsForConstraintsPass` only need to handle aten graph (since it's after AOT export) but if we move it above, it will need to handle torch level graph.
    * The real validation logic for assertion like [`constrain_as_value`](https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_export/constraints.py#L21) is purely implemented as python and the corresponding aten `sym_contrain_range` is no-op and the only purpose of it is to make it possible to survive AOT export (see https://github.com/pytorch/pytorch/pull/103346 for more context). It might worth to move the real assertion logic (in python) so AOT side could see full functional assertion ops before functionalizing them.
  * Keep logic in `_FunctionalizeSideEffectfulOpsPass` but make sure it can reconcile with functionalized ops generated during AOT export.
* There seems to be other functionalization (like https://github.com/pytorch/pytorch/blob/fb8aa721e23fb7b0c751ec31c258cdc3f85a4c31/torch/_functorch/make_functional.py#L348) which is out of scope here?


[ghstack-poisoned]
